File: lib/php/tests/InterfacesTest.php
<?php

declare(strict_types=1);

namespace WgPortal\Client\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use WgPortal\Client\HttpClient;
use WgPortal\Client\InterfacesApi;
use WgPortal\Client\WgPortalException;

class InterfacesTest extends TestCase
{
    private const BASE_URL = 'http://wg-portal.test:8888';

    private function makeApi(array &$queue): InterfacesApi
    {
        $mock = new MockHandler($queue);
        $handlerStack = HandlerStack::create($mock);
        $guzzle = new GuzzleClient(['handler' => $handlerStack]);

        $http = $this->createHttpClient($guzzle);
        return new InterfacesApi($http);
    }

    private function createHttpClient(GuzzleClient $guzzle): HttpClient
    {
        $ref = new \ReflectionClass(HttpClient::class);
        $client = $ref->newInstanceWithoutConstructor();

        $prop = $ref->getProperty('baseUrl');
        $prop->setAccessible(true);
        $prop->setValue($client, self::BASE_URL . '/api/v1');

        $headersProp = $ref->getProperty('headers');
        $headersProp->setAccessible(true);
        $headersProp->setValue($client, [
            'Authorization' => 'Basic dGVzdDp0ZXN0',
        ]);

        $httpProp = $ref->getProperty('http');
        $httpProp->setAccessible(true);
        $httpProp->setValue($client, $guzzle);

        return $client;
    }

    public function testGetAll(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);

        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([
            ['Identifier' => 'wg0', 'Mode' => 'server', 'PrivateKey' => 'k1', 'PublicKey' => 'k2'],
            ['Identifier' => 'wg1', 'Mode' => 'server', 'PrivateKey' => 'k3', 'PublicKey' => 'k4'],
        ]));

        $result = $api->getAll();
        $this->assertCount(2, $result);
        $this->assertEquals('wg0', $result[0]['Identifier']);
    }

    public function testGetAllEmpty(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);

        $queue[] = new Response(200, ['Content-Type' => 'application/json'], '[]');

        $result = $api->getAll();
        $this->assertIsArray($result);
        $this->assertEmpty($result);
    }

    public function testGetById(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);

        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([
            'Identifier' => 'wg0', 'Mode' => 'server', 'PrivateKey' => 'k1', 'PublicKey' => 'k2',
        ]));

        $result = $api->getById('wg0');
        $this->assertEquals('wg0', $result['Identifier']);
        $this->assertEquals('server', $result['Mode']);
    }

    public function testCreate(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);

        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([
            'Identifier' => 'wg0', 'Mode' => 'server', 'PrivateKey' => 'k1', 'PublicKey' => 'k2',
        ]));

        $result = $api->create(['Identifier' => 'wg0', 'Mode' => 'server']);
        $this->assertEquals('wg0', $result['Identifier']);
    }

    public function testDelete(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);

        $queue[] = new Response(204);

        $api->delete('wg0');
        $this->assertTrue(true); // no exception
    }

    public function testGetByIdThrowsOn404(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);

        $queue[] = new Response(
            404,
            ['Content-Type' => 'application/json'],
            '{"Code":404,"Message":"Not Found","Details":"resource not found"}'
        );

        $this->expectException(WgPortalException::class);
        $this->expectExceptionCode(404);

        $api->getById('nonexistent');
    }
}

File: lib/rust/tests/interfaces_test.rs
use mockito::Server;
use wg_portal_client::{WgInterface, WgPortalClient};

fn setup() -> (mockito::ServerGuard, WgPortalClient) {
    let server = Server::new();
    let client = WgPortalClient::new(
        &server.url(),
        "admin@wgportal.local",
        "test-token",
    );
    (server, client)
}

fn fixture_interface() -> serde_json::Value {
    serde_json::json!({
        "Identifier": "wg0",
        "Mode": "server",
        "PrivateKey": "gI6EdUSYvn8ugXOt8QQD6Yc+JyiZxIhp3GInSWRfWGE=",
        "PublicKey": "HIgo9xNzJMWLKASShiTqIybxZ0U3wGLiUeJ1PKf8ykw=",
        "DisplayName": "My Interface",
        "Disabled": false
    })
}

fn fixture_interface_list() -> serde_json::Value {
    serde_json::json!([
        fixture_interface(),
        {
            "Identifier": "wg1",
            "Mode": "server",
            "PrivateKey": "gI6EdUSYvn8ugXOt8QQD6Yc+JyiZxIhp3GInSWRfWGE=",
            "PublicKey": "HIgo9xNzJMWLKASShiTqIybxZ0U3wGLiUeJ1PKf8ykw=",
            "DisplayName": "Second Interface"
        }
    ])
}

#[test]
fn get_all_returns_interfaces() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/interface/all")
        .with_body(fixture_interface_list().to_string())
        .with_header("content-type", "application/json")
        .create();

    let result = client.interfaces().get_all().unwrap();
    assert_eq!(result.len(), 2);
    assert_eq!(result[0].identifier, "wg0");
    mock.assert();
}

#[test]
fn get_all_empty() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/interface/all")
        .with_body("[]")
        .with_header("content-type", "application/json")
        .create();

    let result = client.interfaces().get_all().unwrap();
    assert!(result.is_empty());
    mock.assert();
}

#[test]
fn get_by_id() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/interface/by-id/wg0")
        .with_body(fixture_interface().to_string())
        .with_header("content-type", "application/json")
        .create();

    let result = client.interfaces().get_by_id("wg0").unwrap();
    assert_eq!(result.identifier, "wg0");
    assert_eq!(result.mode, "server");
    mock.assert();
}

#[test]
fn create_interface() {
    let (mut server, client) = setup();
    let mock = server
        .mock("POST", "/api/v1/interface/new")
        .with_body(fixture_interface().to_string())
        .with_header("content-type", "application/json")
        .create();

    let data: WgInterface = serde_json::from_value(fixture_interface()).unwrap();
    let result = client.interfaces().create(&data).unwrap();
    assert_eq!(result.identifier, "wg0");
    mock.assert();
}

#[test]
fn delete_interface() {
    let (mut server, client) = setup();
    let mock = server
        .mock("DELETE", "/api/v1/interface/by-id/wg0")
        .with_status(204)
        .create();

    client.interfaces().delete("wg0").unwrap();
    mock.assert();
}

#[test]
fn get_by_id_throws_on_404() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/interface/by-id/nonexistent")
        .with_status(404)
        .with_body(r#"{"Code":404,"Message":"Not Found","Details":"resource not found"}"#)
        .with_header("content-type", "application/json")
        .create();

    let err = client.interfaces().get_by_id("nonexistent").unwrap_err();
    assert_eq!(err.code, 404);
    mock.assert();
}
